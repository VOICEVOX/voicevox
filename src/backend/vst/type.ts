import { TrackId } from "@/type/preload";

export type Routing = {
  channelMode: "mono" | "stereo";
  channelIndex: Record<TrackId, number>;
};
