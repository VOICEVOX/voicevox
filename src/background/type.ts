import { z } from "zod";
import { engineIdSchema } from "@/type/preload";

// デフォルトエンジン情報
export const defaultEngineInfosEnvSchema = z
  .object({
    uuid: engineIdSchema,
    host: z.string(),
    name: z.string(),
    executionEnabled: z.boolean(),
    executionFilePath: z.string(),
    executionArgs: z.array(z.string()),
    path: z.string().optional(),
  })
  .array();
