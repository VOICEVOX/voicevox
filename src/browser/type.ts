import { IpcIHData } from "@/type/ipc";

export type WorkerToMainMessage = {
  [K in keyof IpcIHData]: {
    type: K;
    return: IpcIHData[K]["return"];
    eventId: string;
  };
}[keyof IpcIHData];

export type MainToWorkerMessage = {
  [K in keyof IpcIHData]: {
    type: K;
    args: IpcIHData[K]["args"];
    eventId: string;
  };
}[keyof IpcIHData];
