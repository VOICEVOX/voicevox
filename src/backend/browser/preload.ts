import { SandboxKey, Sandbox } from "../type/preload";
import { api } from "./sandbox";

const sandbox: Sandbox = api;
// @ts-expect-error readonlyになっているが、初期化処理はここで行うので問題ない
window[SandboxKey] = sandbox;
