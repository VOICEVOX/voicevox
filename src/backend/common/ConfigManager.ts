import semver from "semver";
import AsyncLock from "async-lock";
import {
  AcceptTermsStatus,
  ConfigType,
  EngineId,
  configSchema,
  DefaultStyleId,
  defaultHotkeySettings,
  HotkeySettingType,
  ExperimentalSettingType,
  HotkeyCombination,
} from "@/type/preload";

const lockKey = "save";

const migrations: [string, (store: Record<string, unknown>) => unknown][] = [
  [
    ">=0.13",
    (config) => {
      // acceptTems -> acceptTerms
      const prevIdentifier = "acceptTems";
      const prevValue = config[prevIdentifier] as AcceptTermsStatus | undefined;
      if (prevValue) {
        delete config[prevIdentifier];
        config.acceptTerms = prevValue;
      }

      return config;
    },
  ],
  [
    ">=0.14",
    (config) => {
      // FIXME: できるならEngineManagerからEngineIDを取得したい
      if (import.meta.env.VITE_DEFAULT_ENGINE_INFOS == undefined) {
        throw new Error("VITE_DEFAULT_ENGINE_INFOS == undefined");
      }
      const engineId = EngineId(
        JSON.parse(import.meta.env.VITE_DEFAULT_ENGINE_INFOS)[0].uuid,
      );
      if (engineId == undefined)
        throw new Error("VITE_DEFAULT_ENGINE_INFOS[0].uuid == undefined");
      const prevDefaultStyleIds = config.defaultStyleIds as DefaultStyleId[];
      config.defaultStyleIds = prevDefaultStyleIds.map((defaultStyle) => ({
        engineId,
        speakerUuid: defaultStyle.speakerUuid,
        defaultStyleId: defaultStyle.defaultStyleId,
      }));

      const outputSamplingRate: number = (
        config.savingSetting as { outputSamplingRate: number }
      ).outputSamplingRate;
      const engineSettings: ConfigType["engineSettings"] = {};
      engineSettings[engineId] = {
        useGpu: config.useGpu as boolean,
        outputSamplingRate:
          outputSamplingRate === 24000 ? "engineDefault" : outputSamplingRate,
      };
      config.engineSettings = engineSettings;

      const savingSetting = config.savingSetting as ConfigType["savingSetting"];
      // @ts-expect-error 削除されたパラメータ。
      delete savingSetting.outputSamplingRate;
      config.savingSetting = savingSetting;

      delete config.useGpu;

      return config;
    },
  ],
  [
    ">=0.14.9",
    (config) => {
      // マルチエンジン機能を実験的機能から通常機能に
      const experimentalSetting =
        config.experimentalSetting as ExperimentalSettingType; // FIXME: parseするかasをやめる
      if (
        Object.prototype.hasOwnProperty.call(
          experimentalSetting,
          "enableMultiEngine",
        )
      ) {
        const enableMultiEngine: boolean =
          // @ts-expect-error 削除されたパラメータ。
          config.experimentalSetting.enableMultiEngine;
        config.enableMultiEngine = enableMultiEngine;
        // @ts-expect-error 削除されたパラメータ。
        delete config.experimentalSetting.enableMultiEngine;
      }
    },
  ],
  [
    ">=0.15",
    (config) => {
      // 一つだけ書き出し → 選択音声を書き出し
      const hotkeySettings =
        config.hotkeySettings as ConfigType["hotkeySettings"];
      const newHotkeySettings: ConfigType["hotkeySettings"] =
        hotkeySettings.map((hotkeySetting) => {
          /// @ts-expect-error 名前変更なので合わない。
          if (hotkeySetting.action === "一つだけ書き出し") {
            return {
              ...hotkeySetting,
              action: "選択音声を書き出し",
            };
          }
          return hotkeySetting;
        });
      config.hotkeySettings = newHotkeySettings;

      const toolbarSetting =
        config.toolbarSetting as ConfigType["toolbarSetting"];
      const newToolbarSetting: ConfigType["toolbarSetting"] =
        toolbarSetting.map((toolbarSetting) =>
          // @ts-expect-error 名前変更なので合わない。
          toolbarSetting === "EXPORT_AUDIO_ONE"
            ? "EXPORT_AUDIO_SELECTED"
            : toolbarSetting,
        );
      config.toolbarSetting = newToolbarSetting;

      return config;
    },
  ],
  [
    ">=0.17",
    (config) => {
      // 書き出し先のディレクトリが空文字の場合書き出し先固定を無効化する
      // FIXME: 勝手に書き換えるのは少し不親切なので、ダイアログで書き換えたことを案内する
      const savingSetting = config.savingSetting as ConfigType["savingSetting"];
      if (
        savingSetting.fixedExportEnabled &&
        savingSetting.fixedExportDir === ""
      ) {
        savingSetting.fixedExportEnabled = false;
      }
    },
  ],
  [
    ">=0.19",
    (config) => {
      // ピッチ表示機能の設定をピッチ編集機能に引き継ぐ
      const experimentalSetting =
        config.experimentalSetting as ExperimentalSettingType & {
          showPitchInSongEditor?: boolean; // FIXME: TypeScript 5.4.5ならこの型の結合は不要
        };
      if (
        "showPitchInSongEditor" in experimentalSetting &&
        typeof experimentalSetting.showPitchInSongEditor === "boolean"
      ) {
        experimentalSetting.enablePitchEditInSongEditor =
          experimentalSetting.showPitchInSongEditor;
        delete experimentalSetting.showPitchInSongEditor;
      }
    },
  ],
  [
    ">=0.20",
    (config) => {
      // プロジェクト読み込み → プロジェクトを読み込む
      const hotkeySettings =
        config.hotkeySettings as ConfigType["hotkeySettings"];
      const newHotkeySettings: ConfigType["hotkeySettings"] =
        hotkeySettings.map((hotkeySetting) => {
          /// @ts-expect-error 名前変更なので合わない。
          if (hotkeySetting.action === "プロジェクト読み込み") {
            return {
              ...hotkeySetting,
              action: "プロジェクトを読み込む",
            };
          }
          /// @ts-expect-error 名前変更なので合わない。
          if (hotkeySetting.action === "テキスト読み込む") {
            return {
              ...hotkeySetting,
              action: "テキストを読み込む",
            };
          }
          return hotkeySetting;
        });
      config.hotkeySettings = newHotkeySettings;

      return config;
    },
  ],
];

export type Metadata = {
  __internal__: {
    migrations: {
      version: string;
    };
  };
};

/**
 * 設定管理の基底クラス
 *
 * # ロジックメモ
 * 保存呼び出しのカウンターを用意する。
 * set（save）が呼ばれる度、カウンターをインクリメントし、保存のPromiseをspawnする。
 *
 * 必ず保存されることを保証したい時（アプリ終了時など）は、await ensureSaved()を呼ぶ。
 */
export abstract class BaseConfigManager {
  protected config: ConfigType | undefined;

  private lock = new AsyncLock();

  protected abstract exists(): Promise<boolean>;
  protected abstract load(): Promise<Record<string, unknown> & Metadata>;
  protected abstract save(config: ConfigType & Metadata): Promise<void>;

  protected abstract getAppVersion(): string;

  public reset() {
    this.config = this.getDefaultConfig();
    this._save();
  }

  public async initialize(): Promise<this> {
    if (await this.exists()) {
      const data = await this.load();
      const version = data.__internal__.migrations.version;
      for (const [versionRange, migration] of migrations) {
        if (!semver.satisfies(version, versionRange)) {
          migration(data);
        }
      }
      this.config = this.migrateHotkeySettings(configSchema.parse(data));
      this._save();
    } else {
      this.reset();
    }
    await this.ensureSaved();

    return this;
  }

  public get<K extends keyof ConfigType>(key: K): ConfigType[K] {
    if (!this.config) throw new Error("Config is not initialized");
    return this.config[key];
  }

  public set<K extends keyof ConfigType>(key: K, value: ConfigType[K]) {
    if (!this.config) throw new Error("Config is not initialized");
    this.config[key] = value;
    this._save();
  }

  private _save() {
    this.lock.acquire(lockKey, async () => {
      await this.save({
        ...configSchema.parse({
          ...this.config,
        }),
        __internal__: {
          migrations: {
            version: this.getAppVersion(),
          },
        },
      });
    });
  }

  ensureSaved(): Promise<void> | "alreadySaved" {
    if (!this.lock.isBusy(lockKey)) {
      return "alreadySaved";
    }

    return this._ensureSaved();
  }

  private async _ensureSaved(): Promise<void> {
    // 10秒待っても保存が終わらなかったら諦める
    for (let i = 0; i < 100; i++) {
      // 他のスレッドに処理を譲る
      await new Promise((resolve) => setTimeout(resolve, 100));
      if (!this.lock.isBusy(lockKey)) {
        return;
      }
    }
    throw new Error("Failed to save config");
  }

  private migrateHotkeySettings(data: ConfigType): ConfigType {
    const COMBINATION_IS_NONE = HotkeyCombination("####");
    const loadedHotkeys = structuredClone(data.hotkeySettings);
    const hotkeysWithoutNewCombination = defaultHotkeySettings.map(
      (defaultHotkey) => {
        const loadedHotkey = loadedHotkeys.find(
          (loadedHotkey) => loadedHotkey.action === defaultHotkey.action,
        );
        const hotkeyWithoutCombination: HotkeySettingType = {
          action: defaultHotkey.action,
          combination: COMBINATION_IS_NONE,
        };
        return loadedHotkey || hotkeyWithoutCombination;
      },
    );
    const migratedHotkeys = hotkeysWithoutNewCombination.map((hotkey) => {
      if (hotkey.combination === COMBINATION_IS_NONE) {
        const newHotkey =
          defaultHotkeySettings.find(
            (defaultHotkey) => defaultHotkey.action === hotkey.action,
          ) || hotkey; // ここの find が undefined を返すケースはないが、ts のエラーになるので入れた
        const combinationExists = hotkeysWithoutNewCombination.some(
          (hotkey) => hotkey.combination === newHotkey.combination,
        );
        if (combinationExists) {
          const emptyHotkey: HotkeySettingType = {
            action: newHotkey.action,
            combination: HotkeyCombination(""),
          };
          return emptyHotkey;
        } else {
          return newHotkey;
        }
      } else {
        return hotkey;
      }
    });
    return {
      ...data,
      hotkeySettings: migratedHotkeys,
    };
  }

  protected getDefaultConfig(): ConfigType {
    return configSchema.parse({});
  }
}
