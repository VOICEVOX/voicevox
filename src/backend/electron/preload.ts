import { contextBridge, ipcRenderer } from "electron";
import type { IpcRendererInvoke } from "./ipc";
import {
  ConfigType,
  EngineId,
  Sandbox,
  SandboxKey,
  TextAsset,
} from "@/type/preload";

const ipcRendererInvokeProxy = new Proxy(
  {},
  {
    get:
      (_, channel: string) =>
      (...args: unknown[]) =>
        ipcRenderer.invoke(channel, ...args),
  },
) as IpcRendererInvoke;

const api: Sandbox = {
  getAppInfos: async () => {
    return await ipcRendererInvokeProxy.GET_APP_INFOS();
  },

  getTextAsset: (textType) => {
    return ipcRendererInvokeProxy.GET_TEXT_ASSET(textType) as Promise<
      TextAsset[typeof textType]
    >;
  },

  getAltPortInfos: async () => {
    return await ipcRendererInvokeProxy.GET_ALT_PORT_INFOS();
  },

  showAudioSaveDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_AUDIO_SAVE_DIALOG({
      title,
      defaultPath,
    });
  },

  showTextSaveDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_TEXT_SAVE_DIALOG({ title, defaultPath });
  },

  showSaveDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_SAVE_DIRECTORY_DIALOG({ title });
  },

  showVvppOpenDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_VVPP_OPEN_DIALOG({ title, defaultPath });
  },

  showOpenDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_OPEN_DIRECTORY_DIALOG({ title });
  },

  showProjectSaveDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_PROJECT_SAVE_DIALOG({
      title,
      defaultPath,
    });
  },

  showProjectLoadDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_PROJECT_LOAD_DIALOG({ title });
  },

  showImportFileDialog: ({ title, name, extensions }) => {
    return ipcRendererInvokeProxy.SHOW_IMPORT_FILE_DIALOG({
      title,
      name,
      extensions,
    });
  },

  writeFile: async ({ filePath, buffer }) => {
    return await ipcRendererInvokeProxy.WRITE_FILE({ filePath, buffer });
  },

  readFile: async ({ filePath }) => {
    return await ipcRendererInvokeProxy.READ_FILE({ filePath });
  },

  isAvailableGPUMode: () => {
    return ipcRendererInvokeProxy.IS_AVAILABLE_GPU_MODE();
  },

  isMaximizedWindow: () => {
    return ipcRendererInvokeProxy.IS_MAXIMIZED_WINDOW();
  },

  onReceivedIPCMsg: (listeners) => {
    Object.entries(listeners).forEach(([channel, listener]) => {
      ipcRenderer.on(channel, listener);
    });
  },

  closeWindow: () => {
    void ipcRendererInvokeProxy.CLOSE_WINDOW();
  },

  minimizeWindow: () => {
    void ipcRendererInvokeProxy.MINIMIZE_WINDOW();
  },

  maximizeWindow: () => {
    void ipcRendererInvokeProxy.MAXIMIZE_WINDOW();
  },

  logError: (...params) => {
    console.error(...params);
    // 経緯 https://github.com/VOICEVOX/voicevox/pull/1620#discussion_r1371804569
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "error",
    });
  },

  logWarn: (...params) => {
    console.warn(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "warn",
    });
  },

  logInfo: (...params) => {
    console.info(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "info",
    });
  },

  openLogDirectory: () => {
    void ipcRendererInvokeProxy.OPEN_LOG_DIRECTORY();
  },

  engineInfos: () => {
    return ipcRendererInvokeProxy.ENGINE_INFOS();
  },

  restartEngine: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.RESTART_ENGINE({ engineId });
  },

  openEngineDirectory: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.OPEN_ENGINE_DIRECTORY({ engineId });
  },

  checkFileExists: (file) => {
    return ipcRendererInvokeProxy.CHECK_FILE_EXISTS({ file });
  },

  changePinWindow: () => {
    void ipcRendererInvokeProxy.CHANGE_PIN_WINDOW();
  },

  hotkeySettings: (newData) => {
    return ipcRendererInvokeProxy.HOTKEY_SETTINGS({ newData });
  },

  getDefaultHotkeySettings: async () => {
    return await ipcRendererInvokeProxy.GET_DEFAULT_HOTKEY_SETTINGS();
  },

  getDefaultToolbarSetting: async () => {
    return await ipcRendererInvokeProxy.GET_DEFAULT_TOOLBAR_SETTING();
  },

  setNativeTheme: (source) => {
    void ipcRendererInvokeProxy.SET_NATIVE_THEME(source);
  },

  vuexReady: () => {
    void ipcRendererInvokeProxy.ON_VUEX_READY();
  },

  /**
   * 設定情報を取得する
   */
  getSetting: async (key) => {
    return (await ipcRendererInvokeProxy.GET_SETTING(
      key,
    )) as ConfigType[typeof key];
  },

  /**
   * 設定情報を保存する
   */
  setSetting: async (key, newValue) => {
    return (await ipcRendererInvokeProxy.SET_SETTING(
      key,
      newValue,
    )) as typeof newValue;
  },

  setEngineSetting: async (engineId, engineSetting) => {
    await ipcRendererInvokeProxy.SET_ENGINE_SETTING(engineId, engineSetting);
  },

  installVvppEngine: async (filePath) => {
    return await ipcRendererInvokeProxy.INSTALL_VVPP_ENGINE(filePath);
  },

  uninstallVvppEngine: async (engineId) => {
    return await ipcRendererInvokeProxy.UNINSTALL_VVPP_ENGINE(engineId);
  },

  validateEngineDir: async (engineDir) => {
    return await ipcRendererInvokeProxy.VALIDATE_ENGINE_DIR({ engineDir });
  },

  /**
   * アプリを再読み込みする。
   * 画面以外の情報を刷新する。
   */
  reloadApp: async ({ isMultiEngineOffMode }) => {
    await ipcRendererInvokeProxy.RELOAD_APP({ isMultiEngineOffMode });
  },
};

contextBridge.exposeInMainWorld(SandboxKey, api);
