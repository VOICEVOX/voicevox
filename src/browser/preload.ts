import { SandboxKey } from "../type/preload";
import { api } from "./sandbox";

// @ts-expect-error readonlyになっているが、初期化処理はここで行うので問題ない
window[SandboxKey] = api;
