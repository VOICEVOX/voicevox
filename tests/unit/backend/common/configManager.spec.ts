import pastConfigs from "./pastConfigs";
import { BaseConfigManager } from "@/backend/common/ConfigManager";
import { configSchema } from "@/type/preload";

const configBase = {
  ...configSchema.parse({}),
  __internal__: {
    migrations: {
      version: "999.999.999",
    },
  },
};

class TestConfigManager extends BaseConfigManager {
  getAppVersion() {
    return "999.999.999";
  }

  async exists() {
    throw new Error("mockで実装してください");

    // Unreachableだが、一応booleanを返さないとモックできないので返しておく。
    return false;
  }

  async load() {
    throw new Error("mockで実装してください");

    return {} as ReturnType<BaseConfigManager["load"]>;
  }

  // VitestのmockFn.mock.callsの型のために引数を受け取るようにしている。
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async save(config: Parameters<BaseConfigManager["save"]>[0]) {
    throw new Error("mockで実装してください");
  }
}

it("新規作成できる", async () => {
  vi.spyOn(TestConfigManager.prototype, "exists").mockImplementation(
    async () => false,
  );
  vi.spyOn(TestConfigManager.prototype, "save").mockImplementation(
    async () => undefined,
  );

  const configManager = new TestConfigManager();
  await configManager.initialize();
  expect(configManager).toBeTruthy();
});

it("バージョンが保存される", async () => {
  vi.spyOn(TestConfigManager.prototype, "exists").mockImplementation(
    async () => false,
  );
  const saveSpy = vi
    .spyOn(TestConfigManager.prototype, "save")
    .mockImplementation(async () => undefined);

  const configManager = new TestConfigManager();
  await configManager.initialize();
  await configManager.ensureSaved();
  expect(saveSpy).toHaveBeenCalled();
  const savedData = saveSpy.mock.calls[0][0];
  expect(savedData.__internal__.migrations.version).toBe("999.999.999");
});

for (const [version, data] of pastConfigs) {
  it(`${version}からマイグレーションできる`, async () => {
    vi.spyOn(TestConfigManager.prototype, "exists").mockImplementation(
      async () => true,
    );
    vi.spyOn(TestConfigManager.prototype, "save").mockImplementation(
      async () => undefined,
    );
    vi.spyOn(TestConfigManager.prototype, "load").mockImplementation(
      async () => data,
    );

    const configManager = new TestConfigManager();
    await configManager.initialize();
    expect(configManager).toBeTruthy();
  });
}

it("getできる", async () => {
  vi.spyOn(TestConfigManager.prototype, "exists").mockImplementation(
    async () => true,
  );
  vi.spyOn(TestConfigManager.prototype, "save").mockImplementation(
    async () => undefined,
  );
  vi.spyOn(TestConfigManager.prototype, "load").mockImplementation(
    async () => ({
      ...configBase,
      inheritAudioInfo: false,
    }),
  );

  const configManager = new TestConfigManager();
  await configManager.initialize();
  expect(configManager.get("inheritAudioInfo")).toBe(false);
});

it("setできる", async () => {
  vi.spyOn(TestConfigManager.prototype, "exists").mockImplementation(
    async () => true,
  );
  vi.spyOn(TestConfigManager.prototype, "save").mockImplementation(
    async () => undefined,
  );
  vi.spyOn(TestConfigManager.prototype, "load").mockImplementation(
    async () => ({
      ...configBase,
      inheritAudioInfo: false,
    }),
  );

  const configManager = new TestConfigManager();
  await configManager.initialize();
  configManager.set("inheritAudioInfo", true);
  expect(configManager.get("inheritAudioInfo")).toBe(true);
});
