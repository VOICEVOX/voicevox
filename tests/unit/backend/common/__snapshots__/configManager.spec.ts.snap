// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`0.13.0からマイグレーションできる 1`] = `
{
  "acceptRetrieveTelemetry": "Accepted",
  "acceptTerms": "Accepted",
  "activePointScrollMode": "OFF",
  "confirmedTips": {
    "engineStartedOnAltPort": false,
    "notifyOnGenerate": false,
    "tweakableSliderByScroll": false,
  },
  "currentTheme": "Default",
  "defaultPresetKeys": {},
  "defaultStyleIds": [],
  "editorFont": "default",
  "enableMemoNotation": false,
  "enableMultiEngine": false,
  "enablePreset": false,
  "enableRubyNotation": false,
  "engineSettings": {
    "074fc39e-678b-4c13-8916-ffca8d505d1d": {
      "outputSamplingRate": "engineDefault",
      "useGpu": false,
    },
  },
  "experimentalSetting": {
    "enableInterrogativeUpspeak": false,
    "enableMorphing": false,
    "enableMultiSelect": false,
    "shouldKeepTuningOnTextChange": false,
  },
  "hotkeySettings": [
    {
      "action": "音声書き出し",
      "combination": "Ctrl E",
    },
    {
      "action": "選択音声を書き出し",
      "combination": "E",
    },
    {
      "action": "音声を繋げて書き出し",
      "combination": "",
    },
    {
      "action": "再生/停止",
      "combination": "Space",
    },
    {
      "action": "連続再生/停止",
      "combination": "Shift Space",
    },
    {
      "action": "ｱｸｾﾝﾄ欄を表示",
      "combination": "1",
    },
    {
      "action": "ｲﾝﾄﾈｰｼｮﾝ欄を表示",
      "combination": "2",
    },
    {
      "action": "長さ欄を表示",
      "combination": "3",
    },
    {
      "action": "テキスト欄を追加",
      "combination": "Shift Enter",
    },
    {
      "action": "テキスト欄を複製",
      "combination": "Ctrl D",
    },
    {
      "action": "テキスト欄を削除",
      "combination": "Shift Delete",
    },
    {
      "action": "テキスト欄からフォーカスを外す",
      "combination": "Escape",
    },
    {
      "action": "テキスト欄にフォーカスを戻す",
      "combination": "Enter",
    },
    {
      "action": "元に戻す",
      "combination": "Ctrl Z",
    },
    {
      "action": "やり直す",
      "combination": "Ctrl Y",
    },
    {
      "action": "新規プロジェクト",
      "combination": "Ctrl N",
    },
    {
      "action": "プロジェクトを名前を付けて保存",
      "combination": "Ctrl Shift S",
    },
    {
      "action": "プロジェクトを上書き保存",
      "combination": "Ctrl S",
    },
    {
      "action": "プロジェクトを読み込む",
      "combination": "Ctrl O",
    },
    {
      "action": "テキストを読み込む",
      "combination": "",
    },
    {
      "action": "全体のイントネーションをリセット",
      "combination": "Ctrl G",
    },
    {
      "action": "選択中のアクセント句のイントネーションをリセット",
      "combination": "R",
    },
    {
      "action": "コピー",
      "combination": "Ctrl C",
    },
    {
      "action": "切り取り",
      "combination": "Ctrl X",
    },
    {
      "action": "貼り付け",
      "combination": "Ctrl V",
    },
    {
      "action": "すべて選択",
      "combination": "Ctrl A",
    },
    {
      "action": "選択解除",
      "combination": "",
    },
    {
      "action": "1番目のキャラクターを選択",
      "combination": "Ctrl 1",
    },
    {
      "action": "2番目のキャラクターを選択",
      "combination": "Ctrl 2",
    },
    {
      "action": "3番目のキャラクターを選択",
      "combination": "Ctrl 3",
    },
    {
      "action": "4番目のキャラクターを選択",
      "combination": "Ctrl 4",
    },
    {
      "action": "5番目のキャラクターを選択",
      "combination": "Ctrl 5",
    },
    {
      "action": "6番目のキャラクターを選択",
      "combination": "Ctrl 6",
    },
    {
      "action": "7番目のキャラクターを選択",
      "combination": "Ctrl 7",
    },
    {
      "action": "8番目のキャラクターを選択",
      "combination": "Ctrl 8",
    },
    {
      "action": "9番目のキャラクターを選択",
      "combination": "Ctrl 9",
    },
    {
      "action": "10番目のキャラクターを選択",
      "combination": "Ctrl 0",
    },
  ],
  "inheritAudioInfo": true,
  "playheadPositionDisplayFormat": "MINUTES_SECONDS",
  "presets": {
    "items": {},
    "keys": [],
  },
  "recentlyUsedProjects": [],
  "registeredEngineDirs": [],
  "savingSetting": {
    "audioOutputDevice": "default",
    "avoidOverwrite": false,
    "exportLab": false,
    "exportText": false,
    "fileEncoding": "UTF-8",
    "fileNamePattern": "",
    "fixedExportDir": "",
    "fixedExportEnabled": false,
    "outputStereo": false,
    "songTrackFileNamePattern": "",
  },
  "shouldApplyDefaultPresetOnVoiceChanged": false,
  "showAddAudioItemButton": true,
  "showSingCharacterPortrait": true,
  "showTextLineNumber": false,
  "splitTextWhenPaste": "PERIOD_AND_NEW_LINE",
  "splitterPosition": {},
  "toolbarSetting": [
    "PLAY_CONTINUOUSLY",
    "STOP",
    "EXPORT_AUDIO_SELECTED",
    "EMPTY",
    "UNDO",
    "REDO",
  ],
  "undoableTrackOperations": {
    "panAndGain": true,
    "soloAndMute": true,
  },
  "userCharacterOrder": [
    "7ffcb7ce-00ec-4bdc-82cd-45a8889e43ff",
    "388f246b-8c41-4ac1-8e2d-5d79f3ff56d9",
    "35b2c544-660e-401e-b503-0e14c635303a",
    "3474ee95-c274-47f9-aa1a-8322163d96f1",
    "b1a81618-b27b-40d2-b0ea-27a9ad408c4b",
    "c30dc15a-0992-4f8d-8bb8-ad3b314e6a6f",
    "e5020595-5c5d-4e87-b849-270a518d0dcf",
    "4f51116a-d9ee-4516-925d-21f183e2afad",
    "8eaad775-3119-417e-8cf4-2a10bfd592c8",
    "481fb609-6446-4870-9f46-90c4dd623403",
    "9f3ee141-26ad-437e-97bd-d22298d02ad2",
    "1a17ca16-7ee5-4ea5-b191-2f02ace24d21",
    "67d5d8da-acd7-4207-bb10-b5542d3a663b",
    "0f56c2f2-644c-49c9-8989-94e11f7129d0",
    "044830d2-f23b-44d6-ac0d-b5d733caa900",
  ],
}
`;
